scipy/stats/tests/test_discrete_distns.py

```diff
@@ -627,3 +627,14 @@ def test_betanbinom_kurtosis(self, n, a, b, ref):
         #      return float(fourth_moment/var**2 - 3.)
         assert_allclose(betanbinom.stats(n, a, b, moments="k"),
                         ref, rtol=3e-15)
+
+
+class TestZipf:
+    def test_gh20692(self):
+        # test that int32 data for k generates same output as double
+        k = np.arange(0, 1000)
+        k_int32 = k.astype(np.int32)
+        dist = zipf(9)
+        pmf = dist.pmf(k)
+        pmf_k_int32 = dist.pmf(k_int32)
+        assert_equal(pmf, pmf_k_int32)
```

Comment:
Hey Daniel, I just messed around for a bit and found this case that failed before/passed after a similar patch per your comment on `zipfian`:

<details>

```diff
diff --git a/scipy/stats/_discrete_distns.py b/scipy/stats/_discrete_distns.py
index cda7c4e11..2ee610827 100644
--- a/scipy/stats/_discrete_distns.py
+++ b/scipy/stats/_discrete_distns.py
@@ -1412,7 +1412,8 @@ class zipfian_gen(rv_discrete):
         return 1, n
 
     def _pmf(self, k, a, n):
-        return 1.0 / _gen_harmonic(n, a) / k**a
+        k = k.astype(np.float64)
+        return 1.0 / _gen_harmonic(n, a) * k**-a
 
     def _cdf(self, k, a, n):
         return _gen_harmonic(k, a) / _gen_harmonic(n, a)
diff --git a/scipy/stats/tests/test_discrete_distns.py b/scipy/stats/tests/test_discrete_distns.py
index c5993ebde..bbe9085ce 100644
--- a/scipy/stats/tests/test_discrete_distns.py
+++ b/scipy/stats/tests/test_discrete_distns.py
@@ -627,3 +627,12 @@ class TestBetaNBinom:
         #      return float(fourth_moment/var**2 - 3.)
         assert_allclose(betanbinom.stats(n, a, b, moments="k"),
                         ref, rtol=3e-15)
+
+
+def test_zipfian_gh_20692():
+    k = np.arange(0, 1000)
+    k_int32 = k.astype(np.int32)
+    dist = zipfian(111, 22)
+    pmf = dist.pmf(k)
+    pmf_k_int32 = dist.pmf(k_int32)
+    assert_equal(pmf, pmf_k_int32)
```

</details>

You shouldn't trust my judgement on `stats` stuff, but it does look quite similar in nature as you note. The obtained `pmf` floats are ridiculously tiny, maybe `111` is a bit much, but anyway I just pushed it to error out for integers. It does run lightning fast at least.

Reply:
Added in the last commit.